### DISCUSSION TITLE
Integrating the effort function into LGCP

### DISCUSSION THREAD

**Discussion by gpanunzi** [inlabru-org/inlabru#202] [Q&A]
Good afternoon everyone, I am writing because I am facing an issue with incorporating the effort function as in [Martino et al. (2021)](https://onlinelibrary.wiley.com/doi/full/10.1111/ecog.05843) into the model. Everything was working fine with version 2.7.0 of `inlabru`, so the problem might be related to the mesh (?). Here is my model: 

##EFFORT FUNCTIONS
log_detect_catching = function(x,y,t1, t2)
{
  dens = f.fill(x,y,c_eff_SPDF)
  
  pnorm(dens/qexppnorm(t1, rate = 0.5) +
          t2 - 1,
        log.p = TRUE)
}

log_detect_sightings = function(x,y,t1, t2)
{
  dens = f.fill(x,y,e_eff_SPDF)
  
  pnorm(dens/qexppnorm(t1 - 2, rate = 0.1) +
          t2 - 1,
        log.p = TRUE)
}

###DEFINE MODEL COMPONENTS
cmp = ~ 0 + intercept2(1, mean.linear = 0, prec.linear = 1) +
            SPDE(coordinates, model = spde_s, 
                mapper = bru_mapper(mesh_medit))+
  depth_s(main = depth_val_s, mean.linear = 0, prec.linear = 1) + 
  temp_max(main = temp_max_val, mean.linear = 0, prec.linear = 10) + 
  effort_scale_c(1, prec.linear = 1) + 
  effort_shift_c(1, prec.linear = 0.1) +
  effort_scale_s(1, prec.linear = 1) + 
  effort_shift_s(1, prec.linear = 0.1)
  
formula_c = coordinates ~ intercept2  +
                            SPDE  +
                            depth_s -12 + 
                            temp_max + 
  log_detect_catching(x,y,effort_scale_c,effort_shift_c)

formula_s = coordinates ~ intercept2 +  
                            SPDE  +
                            depth_s -11 +
                            temp_max + 
                            log_detect_sightings(x,y,effort_scale_s,effort_shift_s)

## Run model
lik_c <- like("cp",
              formula = formula_c,
              samplers = boundary, 
              domain = list(coordinates = mesh_medit),
              data = shark1_sp,
              exclude=c("effort_scale_s","effort_shift_s")
              )

lik_s <- like("cp",
              formula = formula_s,
              samplers = boundary, 
              domain = list(coordinates = mesh_medit),
              data = shark2_sp,
              exclude=c("effort_scale_c","effort_shift_c")
              )

fit_nobarr <- bru(components = cmp,
           lik_c,
           lik_s,
           options = list(verbose = F,
                          bru_max_iter = 10))

**I am encountering the following error:
Error in h(simpleError(msg, call)) :
error in evaluating the argument 'obj' in selecting a method for function 'coordinates': object 'x' not found**

e_eff_SPDF and c_eff_SPDF are SpatialPixelsDataFrame.

**Comment by finnlindgren** [comment]
Hi, this is due to a necessary change in how coordinate objects are handled, where it even in the past wasn’t guaranteed that sp coordinate names would be preserved and also extracted as variables available to the component definitions. There’s a note about this in the news for inlabru 2.8.0, see https://cran.r-project.org/web/packages/inlabru/news/news.html

the news note gives a hint for how you can extract the coordinate matrix with coordinates(.data.) and from there you can use [,1] and [,2] to extract the x and y columns you wanted.

Another reason direct coordinate access is discouraged is that it drops crs information; it’s better to pass on the entire geometry (which is easier with sf as it’s available as a tibble “geometry” column).
In your case, I suspect you’re relying on code from old covariate extraction examples that have since been folded into the inlabru code itself, with full sp and sf support in the form of the eval_spatial() methods that are automatically called if the component input is a raster-like object.

**Reply by finnlindgren** [reply]
Looking at the details of your code, it looks like the minimal change would be to replace x,y with a matrix xy, but you can pro ably also replace the f.fill calls with eval_spatial calls, allowing you to pass the entire geometry (sp or sf), including crs info (but reverse argument order compared with f.fill).

**Reply by finnlindgren** [reply]
Also, you need to do `library(sp)`.
